### ISSUE TITLE
Chokidar does not seem to trigger when running in Kubernetes utilizing Azure blob storage as a volume mount

### ISSUE THREAD

**Issue by ainsey11** [paulmillr/chokidar#1203]
**Describe the bug**

When running chokidar inside a container, running inside Kubernetes utilising a volume mount from Azure files, chokidar does not trigger when a new file is placed in the directory.

**Versions (please complete the following information):**
 - Chokidar version 3.5.3
 - Node version 14
 - OS version: node:14-buster-slim

**To Reproduce:**

Steps to reproduce the behavior. Include filename and chokidar config.

Create a volume claim in kubernetes using the azure files driver (https://docs.microsoft.com/en-us/azure/aks/azure-files-csi) 
```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: filedrop
  namespace: test
spec:
  storageClassName: azurefile-storageclass
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 500Gi
```

Mount that inside your container:
```
      volumes:
        - name: filedrop-storage
          persistentVolumeClaim:
            claimName: filedrop
```

launch the container - then see the mount (info redacted for security reasons) :
```
cgroup on /sys/fs/cgroup/blkio type cgroup (ro,nosuid,nodev,noexec,relatime,blkio)
cgroup on /sys/fs/cgroup/rdma type cgroup (ro,nosuid,nodev,noexec,relatime,rdma)
//fe646d3XXXXXXXXXX.file.core.windows.net/pvc-aaafe646d3XXXXXXXXXX on /filedrop type cifs (rw,relatime,vers=3.1.1,cache=strict,username=XXXXXXXXX,uid=0,forceuid,gid=0,forcegid,addr=XXXXXXX,file_mode=0777,dir_mode=0777,soft,persistenthandles,nounix,serverino,mapposix,mfsymlinks,rsize=1048576,wsize=1048576,bsize=1048576,echo_interval=60,actimeo=30)
/dev/sda1 on /etc/hosts type ext4 (rw,relatime,discard)
/dev/sda1 on /dev/termination-log type ext4 (rw,relatime,discard)
/dev/sda1 on /etc/hostname type ext4 (rw,relatime,discard)
/dev/sda1 on /etc/resolv.conf type ext4 (rw,relatime,discard)
shm on /dev/shm type tmpfs (rw,nosuid,nodev,noexec,relatime,size=65536k)
```
observe files placed inside folder:
```
root@importer-8477XXXXXXXpc ls /filedrop/pdf/ -l
total 2870
-rwxrwxrwx 1 root root 2180481 Feb  3 22:10 Test.pdf
```

Chokidar does not trigger an event

Ideally prove a problem by isolating and making it reproducible with a very short sample program, which you could paste here:

```
let w = Chockidar.watch(
            wp.calculatedPath,
            {
                ignored: /(^|[\/\\])\../, // ignore dotfiles
                usePolling:true,
                depth: 1,
                interval: 1000,
                awaitWriteFinish: {
                    stabilityThreshold: 30000,
                    pollInterval: 500
                }
            });
        w.on('add', (file,stats) => this.fileTrigger(file,stats,wp));
```

Most valuable could be one or more test cases for [test.js](https://github.com/paulmillr/chokidar/blob/master/test.js) to demonstrate the problem.

**Expected behavior**
the add event to trigger

**Additional context**
trying to use Azure Files as a file repository to then trigger a service which imports the files into an application. Files are uploaded into Azure over SFTP and then chokidar is to trigger when they have uploaded and begin the import process

**Comment by ainsey11** [comment]
Seems similar to #895 however bumping to the latest version didn't resolve the problem, 
it is quite interesting that Azure does the volume mount as a cifs share, so perhaps it's something funky going on there.

I will try and experiment with bumping to node 16 and if using `useFsEvents` to see if it makes a difference

**Comment by ainsey11** [comment]
put a better test together for readability using the documented example(sorry for lack of paste, I can't copy out of the container shell)

![image](https://user-images.githubusercontent.com/6608275/152441780-66806857-52e8-448d-a58b-27d6ef68e9d7.png)

if I set it running:
![image](https://user-images.githubusercontent.com/6608275/152441709-5cb141a2-ecf2-4675-b47b-375f4c79192a.png)

then upload a file:
![image](https://user-images.githubusercontent.com/6608275/152441846-5bb21624-8540-410f-9785-9ffe33a36bd6.png)

nothing happens, but if I kill and restart my node code it then detects the file:
![image](https://user-images.githubusercontent.com/6608275/152441899-dc2c29b7-941b-4f66-bf7c-da9cc7c201c1.png)

using useFsEvents and polling didn't make a difference sadly

**Comment by ainsey11** [comment]
Done a bit more digging - downgrading to 3.1.1 made no difference (as suggested in #895) 
However, I have found that if I create a file from within the container - my test snippet does trigger:
![image](https://user-images.githubusercontent.com/6608275/152618259-0783166a-f011-46c7-83b2-665d4d22d694.png)

![image](https://user-images.githubusercontent.com/6608275/152618276-a56e437e-824a-4635-84fb-1867fd8ea57d.png)

if I upload through the Azure webUI - still doesn't detect the new file

**Comment by ainsey11** [comment]
Some more notes (sorry for the comment spam)

If I mount the same volume mount into a second container (using the available ReadWriteMany option) and create a file from within that container, the first container picks up the file:

![image](https://user-images.githubusercontent.com/6608275/152618991-ae7dcd9c-b6e3-4929-b79b-33a2d5b381b8.png)
![image](https://user-images.githubusercontent.com/6608275/152619008-071359e8-3af1-41f2-8a85-ca4632827b79.png)

so, it would appear that the issue is caused by the Azure web interface doing something (or not doing something) funky when you upload a file via the azure portal, 

it's as if it doesn't trigger inotify or something of that ilk inside the container,

Going to try and upload a file into the storage account via something such as blobfuse to see if it triggers as expected, if it does then it would appear this is more of a bug on the Azure side, than chockidar itself

**Comment by ainsey11** [comment]
One more update for the evening - I mounted the azure file share into a spare server just as a samba drive and uploaded a file:

![image](https://user-images.githubusercontent.com/6608275/152622732-1a57ac10-3942-4319-b3d7-b54b23ca5f4a.png)

and chokidar didn't detect the new file, if I run `ls` in the container, its there, but chokidar won't detect it until I restart the nodejs code. 

yet, if I create a file in the same share, mounted in a second container, the first container does detect it with chokidar, 

so this appears to be something in the way Azure handles file uploads / writes to the share, causing chokidar not to detect the file. 

My final shout is going to be spawning an sftp server inside a third container with the share mounted and see if it detects the file that way, if so that will work for me (I think) but the behaviour is certainly very strange!

**Comment by ainsey11** [comment]
tl;dr - if the I/O comes from inside K8's, you're fine and everything works, if it comes from an external system, the inotify event doesn't get forwarded to the cluster.

Some more information, if I run inotifywatch inside the container and do a simple touch from another container inside k8's, I see inotify events/activity,

If I do the upload via the webUI, or via a cifs mount from the windows VM (like in my last comment) there's no inotify activity,

![image](https://user-images.githubusercontent.com/6608275/152623451-318ea4f1-69e4-44b9-a1cd-25bd29af4116.png)

which leads me to think that K8's is somehow intercepting the I/O when the file is "touched" inside the second container, and notifying the first container (with chokidar in) to the write, but this doesn't get triggered when the write is outside of K8's visibility. I can't find any evidence of that on the internet though, so purely a hunch, but if my theory is true, if I do a volume mount to a third container on my home kube cluster to the azure share and do a write, there should be no inotify events triggered on the azure side of things,

I did the above in my local K8's cluster, so it would appear my theory of K8's incertepting the I/O within the cluster rings true, I/O from Azure files, OR another server mounting the Azure files over CIFS OR another K8's cluster does not trigger an inotify event inside the container mounting the Azure Files storage,

I can see from various sources that the reccomended way is to use Blob Storage over Azure Files, however in my scenario that won't work due to needing a read write many situation,

Every route I head down leads me to this issue: https://github.com/Azure/azure-webjobs-sdk-extensions/issues/14 where the Azure Event Grid does not support Azure files, It looks like Azure uses event grid under the hood to inform the K8's cluster of the I/O, lack of Azure Files support in Event Grid backs up my theory even more,

Thankfully in my use case, if I run an SFTP server inside the cluster, the inotify event triggers as expected due to K8's seeing the I/O natively, it works and everything is as expected, I was planning on using the native SFTP connection into the Azure storage account when it's generally available, but it would appear if I were to use that, I will hit the same problem with the I/O events due to the write taking place outside the cluster,

The only "workaround" I can think of would be to use the event grid to trigger an API on the file system monitor inside the contanier, but.. as event grid doesn't support Azure Files - you're SOL! 

Very much looking like a Kube / Azure bug now over a chokidar bug so will mark this as closed, I don't think there is anything chokidar can to in this scenario.

@paulmillr - do you think this may be worth adding this as a "known issue" into the docs? Just thinking if another poor soul hits the same issue in the future, may save them some time bug hunting

**Comment by paulmillr** [comment]
@ainsey11 thanks for the investigation. I think google will help people to find this issue, you've added more than enough search tags here.

Network bugs are well-known, and Chokidar was never really aimed to serve as "network disk watcher" solution, or "cloud watcher", or whatever. This is just a side feature that some people discovered.
